tutorials/11-probabilistic-pca/11-probabilistic-pca.jmd

```diff
@@ -0,0 +1,385 @@
+rename!(df_pre, Symbol.( ["z"*string(i) for i in collect(1:n_genes)]))
+df_pre[!,:cell] = 1:n_cells
+
+DataFrames.stack(df_pre, 1:n_genes) |> @vlplot(:rect, "cell:o", "variable:o", color=:value)
+
+df_pre[!,:type] = repeat([1, 2], inner = n_cells÷2)
+df_pre[!,:ard1] = df_pre[:, alpha_indices[1]]
+df_pre[!,:ard2] = df_pre[:, alpha_indices[2]]
+df_pre |>  @vlplot(:point, x=:ard1, y=:ard2, color="type:n")
+```
+
+## Batch effects
+
+Finally, a very common issue to address in biological data is [batch effects](https://en.wikipedia.org/wiki/Batch_effec).
+A batch effect occurs when non-biological factors in an experiment cause changes in the data produced by the experiment.
+As an example, we will look at Fisher's famous Iris data set.
+
+The data set consists of 50 samples each from three species of Iris (Iris setosa, Iris virginica and Iris versicolor).
+Four features were measured from each sample: the length and the width of the sepals and petals, in centimeters.
+
+```julia
+# Example data set - generate synthetic gene expression data
+using RDatasets
+data = dataset("datasets", "iris")
+species = data[!, "Species"]
+dat = data[!, 1:4]
+n, d = size(dat)
+```
+
+
+```julia
+ppca = pPCA(dat)
+
+# Here we use a different samples, we don't always have to use NUTS:
+# Hamiltonian Monte Carlo (HMC) sampler parameters
+n_iterations = 500
+ϵ = 0.05
+τ = 10
+
+chain = sample(ppca, HMC(ϵ, τ), n_iterations)
+
+# Extract paramter estimates for plotting - mean of posterior
+w = permutedims(reshape(mean(group(chain, :w))[:,2], (d,d)))
+z = permutedims(reshape(mean(group(chain, :z))[:,2], (d, n)))'
+mu = mean(group(chain, :mean))[:,2]
+
+X = w * z
+# X = w * z .+ mu
+
+df_rec = DataFrame(X', :auto)
+df_rec[!,:species] = species
+DataFrames.stack(df_rec, 1:d) |> @vlplot(:rect, "species:o", "variable:o", color=:value)
+
+df_iris = DataFrame(z', :auto)
+rename!(df_iris, Symbol.( ["z"*string(i) for i in collect(1:d)]))
+df_iris[!,:sample] = 1:n
+df_iris[!,:species] = species
+
+df_iris |>  @vlplot(:point, x=:z1, y=:z2, color="species:n")
+```
+
+We now simulate a batch effect; imagine the person taking the measurement uses two different rulers and they are slightly off.
+Again, in practice there are many different reasons for why batch effects occur and it is not always clear what is really at the basis of them,
+nor can they always be tackled via the experimental setup. So we need methods to deal with them.
+
+```julia
+## Introduce batch effect
+batch = rand(Binomial(1, 0.5), 150)
+effect = rand(Normal(2.4, 0.6), 150)
+batch_dat = dat .+ batch .* effect
+
+ppca_batch = pPCA(batch_dat)
+chain = sample(ppca_batch, HMC(ϵ, τ), 500)
+describe(chain)[1]
+
+z = permutedims(reshape(mean(group(chain, :z))[:,2], (d, n)))'
+df_pre = DataFrame(z')
+rename!(df_pre, Symbol.( ["z"*string(i) for i in collect(1:d)]))
+df_pre[!,:sample] = 1:n
+df_pre[!,:species] = species
+df_pre[!,:batch] = batch
+
+df_pre |>  @vlplot(:point, x=:z1, y=:z2, color="species:n", shape=:batch)
```

Comment:
Okay, for this graph, you should note why it's different from the one above -- what did the batch effect do? To me, it looks like the classification problem got a lot harder. Make that explicit!